From 3f8d69f749fdab3ba82818848da0c3f67e1cabaa Mon Sep 17 00:00:00 2001
From: Vignesh C <quic_vignc@quicinc.com>
Date: Wed, 12 Feb 2025 00:01:38 +0530
Subject: [PATCH 095/141] wifi: ath12k: Split scan request for split band
 device

When supplicant initiates scan req for split band device,
mac80211 generates an single scan req to driver with all the
channel frequencies. This is because mac80211 categorize the
the frequencies based on band only.

Driver checks the first freq from the channel list and initiates
scan to firmware with all the frequencies. Firmware rejects the freq
which are not supported. As a result of this, Supplicant receives
the scan results from only one device and associates on one link.

Fix this by splitting the scan request in driver and perform
multiple scans, send completion notification to mac80211 after
split scans are done.

Signed-off-by: Vignesh C <quic_vignc@quicinc.com>
Signed-off-by: Rameshkumar Sundaram <rameshkumar.sundaram@oss.qualcomm.com>
Signed-off-by: Raj Kumar Bhagat <quic_rajkbhag@quicinc.com>
Signed-off-by: Jeff Johnson <jeff.johnson@oss.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/core.h |   1 +
 drivers/net/wireless/ath/ath12k/mac.c  | 128 ++++++++++++++++++++-----
 2 files changed, 107 insertions(+), 22 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.h b/drivers/net/wireless/ath/ath12k/core.h
index 97db1a34e7c4..947f20ae2d62 100644
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -341,6 +341,7 @@ struct ath12k_link_vif {
 	u8 current_cntdown_counter;
 	struct ath12k_link_stats link_stats;
 	spinlock_t link_stats_lock; /* Protects updates to link_stats */
+	bool is_scan_vif;
 };
 
 struct ath12k_vif {
diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 250ec23918ed..a1a06e971ee6 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -3980,7 +3980,7 @@ static struct ath12k_link_vif *ath12k_mac_assign_link_vif(struct ath12k_hw *ah,
 	if (arvif)
 		return arvif;
 
-	if (!vif->valid_links) {
+	if (!vif->valid_links && !ahvif->links_map) {
 		/* Use deflink for Non-ML VIFs and mark the link id as 0
 		 */
 		link_id = 0;
@@ -4736,6 +4736,37 @@ static void ath12k_scan_timeout_work(struct work_struct *work)
 	wiphy_unlock(ath12k_ar_to_hw(ar)->wiphy);
 }
 
+static void ath12k_mac_scan_send_complete(struct ath12k *ar,
+					  struct cfg80211_scan_info info)
+{
+	struct ath12k *partner_ar;
+	struct ath12k_pdev *pdev;
+	struct ath12k_base *ab;
+	struct ath12k_hw_group *ag = ar->ab->ag;
+	bool send_completion = true;
+	struct ath12k_hw *ah = ar->ah;
+	int i, j;
+
+	for (i = 0; i < ag->num_devices; i++) {
+		ab = ag->ab[i];
+		for (j = 0; j < ab->num_radios; j++) {
+			pdev = &ab->pdevs[j];
+			partner_ar = pdev->ar;
+
+			if (!partner_ar || partner_ar == ar)
+				continue;
+			if (partner_ar->scan.state == ATH12K_SCAN_RUNNING) {
+				send_completion = false;
+				break;
+			}
+		}
+		if (!send_completion)
+			break;
+	}
+	if (send_completion)
+		ieee80211_scan_completed(ah->hw, &info);
+}
+
 static void ath12k_scan_vdev_clean_work(struct wiphy *wiphy, struct wiphy_work *work)
 {
 	struct ath12k *ar = container_of(work, struct ath12k,
@@ -4774,7 +4805,7 @@ static void ath12k_scan_vdev_clean_work(struct wiphy *wiphy, struct wiphy_work *
 				    ATH12K_SCAN_STARTING)),
 		};
 
-		ieee80211_scan_completed(ar->ah->hw, &info);
+		ath12k_mac_scan_send_complete(ar, info);
 	}
 
 	ar->scan.state = ATH12K_SCAN_IDLE;
@@ -4976,18 +5007,25 @@ ath12k_mac_find_link_id_by_ar(struct ath12k_vif *ahvif, struct ath12k *ar)
 			return link_id;
 	}
 
+	/* When it comes to STA vdev, make sure to pick the next available link.
+	 * There are cases where single scan req needs to be split in driver
+	 * and initiate separate scan requests to firmware based on device.
+	 */
+	if (ahvif->vdev_type == WMI_VDEV_TYPE_STA)
+		return ffs(~ahvif->links_map) - 1;
 	/* input ar is not assigned to any of the links of ML VIF, use scan
 	 * link (15) for scan vdev creation.
 	 */
 	return ATH12K_DEFAULT_SCAN_LINK;
 }
 
-static int ath12k_mac_op_hw_scan(struct ieee80211_hw *hw,
-				 struct ieee80211_vif *vif,
-				 struct ieee80211_scan_request *hw_req)
+static int ath12k_mac_initiate_hw_scan(struct ieee80211_hw *hw,
+				       struct ieee80211_vif *vif,
+				       struct ieee80211_scan_request *hw_req,
+				       struct ath12k *ar,
+				       u8 from_index, u8 to_index)
 {
 	struct ath12k_hw *ah = ath12k_hw_to_ah(hw);
-	struct ath12k *ar;
 	struct ath12k_vif *ahvif = ath12k_vif_to_ahvif(vif);
 	struct ath12k_link_vif *arvif;
 	struct cfg80211_scan_request *req = &hw_req->req;
@@ -4996,18 +5034,12 @@ static int ath12k_mac_op_hw_scan(struct ieee80211_hw *hw,
 	int ret;
 	int i;
 	bool create = true;
+	u8 n_channels = to_index - from_index;
 
 	lockdep_assert_wiphy(hw->wiphy);
 
 	arvif = &ahvif->deflink;
 
-	/* Since the targeted scan device could depend on the frequency
-	 * requested in the hw_req, select the corresponding radio
-	 */
-	ar = ath12k_mac_select_scan_device(hw, vif, hw_req->req.channels[0]->center_freq);
-	if (!ar)
-		return -EINVAL;
-
 	/* check if any of the links of ML VIF is already started on
 	 * radio(ar) correpsondig to given scan frequency and use it,
 	 * if not use scan link (link 15) for scan purpose.
@@ -5048,8 +5080,10 @@ static int ath12k_mac_op_hw_scan(struct ieee80211_hw *hw,
 		 */
 		arvif = ath12k_mac_assign_link_vif(ah, vif, link_id);
 
+		arvif->is_scan_vif = true;
 		ret = ath12k_mac_vdev_create(ar, arvif);
 		if (ret) {
+			ath12k_mac_unassign_link_vif(arvif);
 			ath12k_warn(ar->ab, "unable to create scan vdev %d\n", ret);
 			return -EINVAL;
 		}
@@ -5103,8 +5137,8 @@ static int ath12k_mac_op_hw_scan(struct ieee80211_hw *hw,
 		arg->scan_f_passive = 1;
 	}
 
-	if (req->n_channels) {
-		arg->num_chan = req->n_channels;
+	if (n_channels) {
+		arg->num_chan = n_channels;
 		arg->chan_list = kcalloc(arg->num_chan, sizeof(*arg->chan_list),
 					 GFP_KERNEL);
 		if (!arg->chan_list) {
@@ -5113,7 +5147,7 @@ static int ath12k_mac_op_hw_scan(struct ieee80211_hw *hw,
 		}
 
 		for (i = 0; i < arg->num_chan; i++)
-			arg->chan_list[i] = req->channels[i]->center_freq;
+			arg->chan_list[i] = req->channels[i + from_index]->center_freq;
 	}
 
 	ret = ath12k_start_scan(ar, arg);
@@ -5158,6 +5192,48 @@ static int ath12k_mac_op_hw_scan(struct ieee80211_hw *hw,
 	return ret;
 }
 
+static int ath12k_mac_op_hw_scan(struct ieee80211_hw *hw,
+				 struct ieee80211_vif *vif,
+				 struct ieee80211_scan_request *hw_req)
+{
+	struct ath12k *ar, *prev_ar;
+	int i, from_index, to_index;
+
+	lockdep_assert_wiphy(hw->wiphy);
+	/* Since the targeted scan device could depend on the frequency
+	 * requested in the hw_req, select the corresponding radio
+	 */
+	prev_ar = ath12k_mac_select_scan_device(hw, vif,
+						hw_req->req.channels[0]->center_freq);
+	if (!prev_ar) {
+		ath12k_warn(ar->ab, "unable to select device for scan\n");
+		return -EINVAL;
+	}
+
+	/* NOTE: There could be 5G low/high channels as mac80211 sees
+	 * it as an single band. In that case split the hw request and
+	 * perform multiple scans
+	 */
+	from_index = 0;
+	for (i = 1; i < hw_req->req.n_channels; i++) {
+		ar = ath12k_mac_select_scan_device(hw, vif,
+						   hw_req->req.channels[i]->center_freq);
+		if (!ar) {
+			ath12k_warn(ar->ab, "unable to select device for scan\n");
+			return -EINVAL;
+		}
+		if (prev_ar == ar)
+			continue;
+
+		to_index = i;
+		ath12k_mac_initiate_hw_scan(hw, vif, hw_req, prev_ar,
+					    from_index, to_index);
+		from_index = to_index;
+		prev_ar = ar;
+	}
+	return ath12k_mac_initiate_hw_scan(hw, vif, hw_req, prev_ar, from_index, i);
+}
+
 static void ath12k_mac_op_cancel_hw_scan(struct ieee80211_hw *hw,
 					 struct ieee80211_vif *vif)
 {
@@ -8865,12 +8941,12 @@ int ath12k_mac_vdev_create(struct ath12k *ar, struct ath12k_link_vif *arvif)
 	struct ieee80211_vif *vif = ath12k_ahvif_to_vif(ahvif);
 	struct ath12k_wmi_vdev_create_arg vdev_arg = {0};
 	struct ath12k_wmi_peer_create_arg peer_param = {0};
-	struct ieee80211_bss_conf *link_conf;
+	struct ieee80211_bss_conf *link_conf = NULL;
 	u32 param_id, param_value;
 	u16 nss;
 	int i;
 	int ret, vdev_id;
-	u8 link_id;
+	u8 link_id, link_addr[ETH_ALEN];
 
 	lockdep_assert_wiphy(hw->wiphy);
 
@@ -8895,13 +8971,18 @@ int ath12k_mac_vdev_create(struct ath12k *ar, struct ath12k_link_vif *arvif)
 	}
 
 	link_conf = wiphy_dereference(hw->wiphy, vif->link_conf[link_id]);
-	if (!link_conf) {
+	if (!link_conf && !arvif->is_scan_vif) {
 		ath12k_warn(ar->ab, "unable to access bss link conf in vdev create for vif %pM link %u\n",
 			    vif->addr, arvif->link_id);
 		return -ENOLINK;
 	}
 
-	memcpy(arvif->bssid, link_conf->addr, ETH_ALEN);
+	if (link_conf) {
+		memcpy(arvif->bssid, link_conf->addr, ETH_ALEN);
+	} else {
+		eth_random_addr(link_addr);
+		memcpy(arvif->bssid, link_addr, ETH_ALEN);
+	}
 
 	arvif->ar = ar;
 	vdev_id = __ffs64(ab->free_vdev_map);
@@ -9057,7 +9138,10 @@ int ath12k_mac_vdev_create(struct ath12k *ar, struct ath12k_link_vif *arvif)
 		break;
 	}
 
-	arvif->txpower = link_conf->txpower;
+	if (link_conf)
+		arvif->txpower = link_conf->txpower;
+	else
+		arvif->txpower = NL80211_TX_POWER_AUTOMATIC;
 	ret = ath12k_mac_txpower_recalc(ar);
 	if (ret)
 		goto err_peer_del;
@@ -11691,7 +11775,7 @@ static const struct ieee80211_ops ath12k_ops = {
 	.hw_scan                        = ath12k_mac_op_hw_scan,
 	.cancel_hw_scan                 = ath12k_mac_op_cancel_hw_scan,
 	.set_key                        = ath12k_mac_op_set_key,
-	.set_rekey_data	                = ath12k_mac_op_set_rekey_data,
+	.set_rekey_data                 = ath12k_mac_op_set_rekey_data,
 	.sta_state                      = ath12k_mac_op_sta_state,
 	.sta_set_txpwr			= ath12k_mac_op_sta_set_txpwr,
 	.link_sta_rc_update		= ath12k_mac_op_link_sta_rc_update,
-- 
